### PR TITLE
🐛 fix bus error on smaller readonly file in unix

### LINKED ISSUE #28245: Mmap bus error
```julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0-beta2.84 (2018-07-23 19:53 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 10002fcb5e* (0 days old master)
|__/                   |  x86_64-apple-darwin16.7.0

julia> using Mmap

julia> a = Mmap.mmap("foo2", Vector{NTuple{1, Int8}}, 1, 0)
1-element Array{Tuple{Int8},1}:
Bus error: 10
```
Also happens for simple structs like `struct Wrapper; x::Int; end`. I plan on taking a look at this to see what's going on, but if anyone else has any ideas, please share in the mean time.

### PR DESCRIPTION
**Pull request by xgdgsc**
fix https://github.com/JuliaLang/julia/issues/28245 in unix

### CODE REVIEW

File: stdlib/Mmap/src/Mmap.jl [hunk 1]
```diff
@@ -212,6 +212,7 @@ function mmap(io::IO,
     @static if Sys.isunix()
         prot, flags, iswrite = settings(file_desc, shared)
         iswrite && grow && grow!(io, offset, len)
+        !iswrite && mmaplen > filesize(io) && throw(ArgumentError("unable to mmap file size smaller than $mmaplen due to read-only permissions. Please check your dims"))
```
Comment (vtjnash) [inline stdlib/Mmap/src/Mmap.jl:215]:
```suggestion
        szfile = convert(Csize_t, len + offset)
        !iswrite && szfile > filesize(io) && throw(ArgumentError("unable to increase file size to $szfile due to read-only permissions"))
```
(stealing from windows, though we can potentially DRY some of the code here)

File: stdlib/Mmap/src/Mmap.jl [hunk 2]
```diff
@@ -212,6 +212,8 @@ function mmap(io::IO,
     @static if Sys.isunix()
         prot, flags, iswrite = settings(file_desc, shared)
         iswrite && grow && grow!(io, offset, len)
+        szfile = convert(Csize_t, len + offset)
+        !iswrite && szfile > filesize(io) && throw(ArgumentError("unable to increase file size to $szfile due to read-only permissions"))
```
Comment (vtjnash) [inline stdlib/Mmap/src/Mmap.jl:216]:
```suggestion
        szfile = convert(Csize_t, len + offset)
        if iswrite && grow
            grow!(io, offset, len)
        elseif szfile > filesize(io)
            throw(ArgumentError("unable to increase file size to $szfile due to read-only permissions"))
        end
```

Can you make this same change to the Windows code (to respect the `grow` parameter):
```julia
        (readonly || !grow) && szfile > filesize(io) && throw(ArgumentError("unable to increase file size to $szfile due to read-only permissions"))
```

**Comment by giordano** [comment]
Tests are extremely unhappy on almost all platforms

**Comment by vtjnash** [comment]
We may be able to emulate (aka greatly accelerate) the filesize test by doing a seekend instead and checking the position?

**Comment by xgdgsc** [comment]
As `grow!` does nothing to Anonymous. I did the above commit. What do you think of the macOS failure? I looked at the c code and have no idea. Would it be fine to keep the old behaviour for macOS?

File: stdlib/Mmap/src/Mmap.jl [hunk 3]
```diff
@@ -208,18 +208,39 @@ function mmap(io::IO,
     mmaplen = (offset - offset_page) + len
 
     file_desc = gethandle(io)
+    szfile = convert(Csize_t, len + offset)
+    requestedSizeLarger = false
+    if io isa Mmap.Anonymous
+    else
```
Comment (StefanKarpinski) [inline stdlib/Mmap/src/Mmap.jl:214]:
```suggestion
    if !(io isa Mmap.Anonymous)
```

**Comment by KristofferC** [comment]
@vtjnash, could you take another look at this?

**Comment by vtjnash** [comment]
Tests are still failing, so there is not much more I can really do here for review. My previous suggestion for a fix seems to have been unsuccessful, since the seekend is failing.
```
Error in testset SharedArrays:
Error During Test at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-x64-2.0/build/default-macmini-x64-2-0/julialang/julia-master/julia-926f1dde71/share/julia/test/testdefs.jl:21
  Got exception outside of a @test
  LoadError: IOError: filesize: Illegal seek for <fd 21>
  Stacktrace:
    [1] filesize(s::IOStream)
      @ Base ./iostream.jl:226
```

**Comment by vtjnash** [comment]
Actually, IIUC, this is a very weird file on macOS (from shm_open). I can read and write, but not seek the fd. And stat is a bit awkward, as Julia cannot return it:
```
julia> Base.StatStruct(stat_buf)
StatStruct for ""
   size: 4096 bytes
 device: 0
  inode: 0
   mode: 0o000600 (-rw-------)
  nlink: 0
    uid: 501 (jameson)
    gid: 20 (staff)
   rdev: 0
  blksz: 0
 blocks: 0
  mtime: 1969-12-31T19:00:00-0500 (19401 days ago)
  ctime: 1969-12-31T19:00:00-0500 (19401 days ago)

julia> stat(s)
ERROR: stat returned zero type for a valid path
```
implemented here: https://github.com/apple/darwin-xnu/blob/a1babec6b135d1f35b2590a1990af3c5c5393479/bsd/kern/posix_shm.c#L695-L700

This seems to be an implementation quality issue with `IOStream`, where we attempt to use seekend as a fast-path for `filesize`, but we should fall back to attempting `fstat` when that fails.

Unrelatedly, there appears to be a minor issue in SharedArrays forgetting to close the `fd_mem` handle in this functions caller.

**Comment by xgdgsc** [comment]
Not sure whether the InteractiveUtils test failure is related. Would it be fine to fix it only for linux first like this and wait for macOS solution later? Guess not many would encounter this bug on macOS.

**Comment by DilumAluthge** [comment]
CI is all green on this.

@vtjnash Is this good to merge?

**Comment by mkitti** [comment]
As reported in https://github.com/JuliaLang/julia/issues/54203 this fix introduced another issue since `filesize` seeks which may not be appropriate for certain kinds of `IO`.

**Pull request merged**